File: src/main/java/systems/conduit/main/core/events/EventManager.java
package systems.conduit.main.core.events;

import systems.conduit.main.Conduit;
import systems.conduit.main.core.events.annotations.Listener;
import systems.conduit.main.core.events.types.EventType;
import systems.conduit.main.core.plugin.Plugin;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.util.*;

public class EventManager {

    @SuppressWarnings("unchecked")
    private static Optional<Class<? extends EventType>> findEventTypeFromListenerMethod(Method method) {
        Parameter[] params = method.getParameters();
        if (params.length < 1) return Optional.empty();  // We have to have at least one parameter to check.

        // Check to make sure that this parameter is of the EventType type.
        Parameter param = params[0];
        if (!EventType.class.isAssignableFrom(param.getType())) return Optional.empty();

        return Optional.of((Class<? extends EventType>) param.getType());
    }

    public void registerEventClass(Plugin plugin, EventListener listener) {
        // Attempt to create an instance of the listener.
        if (listener == null) return;

        Method[] methods = listener.getClass().getMethods();
        for (Method method : methods) {
            // If this method is not annotated, then we don't care about it.
            if (!method.isAnnotationPresent(Listener.class)) continue;
            Listener annotation = method.getAnnotation(Listener.class);

            registerHandler(plugin, method, listener);
        }
    }

    private void registerHandler(Plugin plugin, Method method, EventListener listener) {
        // Attempt to get the type of event that this method is handling.
        Optional<Class<? extends EventType>> eventType = findEventTypeFromListenerMethod(method);
        if (!eventType.isPresent()) return;  // Skip this method as it doesn't look like it actually handles anything.

        int eventId = EventTypeRegistry.getEventMappings().indexOf(eventType.get());
        if (eventId == -1) {
            Conduit.getLogger().error("Invalid event type register: " + eventType);
            return;
        }

        if (!plugin.getEvents().containsKey(eventId)) {
            // No currently registered events on this.
            Map<EventListener, List<Method>> handlers = new HashMap<>();
            List<Method> methods = new ArrayList<>();
            methods.add(method);
            handlers.put(listener, methods);
            plugin.getEvents().put(eventId, handlers);
            return;
        }

        if (plugin.getEvents().get(eventId).containsKey(listener)) {
            // We have this class already!
            plugin.getEvents().get(eventId).get(listener).add(method);
            return;
        }

        plugin.getEvents().get(eventId).put(listener, Collections.singletonList(method));
    }

    public void dispatchEvent(EventType eventType) {
        int eventId = EventTypeRegistry.getEventMappings().indexOf(eventType.getClass());
        if (eventId == -1) {
            Conduit.getLogger().error("Cannot dispatch invalid event type: " + eventType.getClass());
            return;
        }

        // TODO: Maybe we should pre-process this garbage?
        Conduit.getPluginManager().getPlugins().forEach(plugin -> plugin.getEvents().getOrDefault(eventId, new HashMap<>()).entrySet().stream().sorted((o1, o2) -> {
            Listener firstAnnotation = o1.getKey().getClass().getAnnotation(Listener.class);
            Listener secondAnnotation = o2.getKey().getClass().getAnnotation(Listener.class);
            if (firstAnnotation == null || secondAnnotation == null) return 0;

            return Integer.compare(firstAnnotation.priority(), secondAnnotation.priority());
        }).forEach((entry) -> entry.getValue().forEach(method -> {
            try {
                method.invoke(entry.getKey(), eventType);
            } catch (IllegalAccessException | InvocationTargetException e) {
                Conduit.getLogger().error("Failed to execute plugin event handler method");
                e.printStackTrace();
            }
        })));
    }
}

File: src/main/kotlin/systems/conduit/core/events/EventManager.kt
package systems.conduit.core.events

import systems.conduit.core.Conduit
import systems.conduit.core.events.annotations.Listener
import systems.conduit.core.events.types.EventType
import systems.conduit.core.events.EventTypeRegistry
import systems.conduit.core.plugin.Plugin
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

class EventManager {

    fun registerEventClass(plugin: Plugin, clazz: Class<out EventListener>) {
        // Attempt to create an instance of the listener.
        var listener: EventListener? = null
        try {
            listener = clazz.getConstructor().newInstance()
        } catch (e: NoSuchMethodException) {
            Conduit.logger.error("Failed to create instance of event listener: " + clazz.name)
            e.printStackTrace()
        } catch (e: IllegalAccessException) {
            Conduit.logger.error("Failed to create instance of event listener: " + clazz.name)
            e.printStackTrace()
        } catch (e: InstantiationException) {
            Conduit.logger.error("Failed to create instance of event listener: " + clazz.name)
            e.printStackTrace()
        } catch (e: InvocationTargetException) {
            Conduit.logger.error("Failed to create instance of event listener: " + clazz.name)
            e.printStackTrace()
        }
        if (listener == null) return
        val methods = clazz.methods
        for (method in methods) {
            // If this method is not annotated, then we don't care about it.
            if (!method.isAnnotationPresent(Listener::class.java)) continue
            // TODO do we need this?
            val annotation = method.getAnnotation(Listener::class.java)
            registerHandler(plugin, method, listener)
        }
    }

    private fun registerHandler(plugin: Plugin, method: Method, listener: EventListener) {
        // Attempt to get the type of event that this method is handling
        // Skip this method if it doesn't look like it actually handles anythin.
        val eventType = findEventTypeFromListenerMethod(method) ?: return
        val eventId: Int = EventTypeRegistry.getEventMappings().indexOf(eventType)
        if (eventId == -1) {
            Conduit.logger.error("Invalid event type register: $eventType")
            return
        }
        if (!plugin.events.containsKey(eventId)) {
            // No currently registered events on this.
            val handlers: MutableMap<EventListener, MutableList<Method>> = HashMap()
            val methods: MutableList<Method> = ArrayList()
            methods.add(method)
            handlers[listener] = methods
            plugin.events[eventId] = handlers
            return
        }
        val current = plugin.events[eventId]!!
        HashMap(current).forEach { (key: EventListener, value: MutableList<Method>) ->
            if (key.javaClass == listener.javaClass) {
                // If we found another handler that matches this one, then add to it.
                value.add(method)
                current[key] = value
            }
        }
        plugin.events[eventId] = current
    }

    fun dispatchEvent(eventType: EventType) {
        val eventId: Int = EventTypeRegistry.getEventMappings().indexOf(eventType.javaClass)
        if (eventId == -1) {
            Conduit.logger.error("Cannot dispatch invalid event type: " + eventType.javaClass)
            return
        }

        // TODO: Maybe we should pre-process this garbage?
        Conduit.pluginManager.plugins.forEach { plugin ->
            plugin.events.getOrDefault(eventId, HashMap()).entries.stream().sorted { o1: Map.Entry<EventListener, List<Method?>?>, o2: Map.Entry<EventListener, List<Method?>?> ->
                val firstAnnotation = o1.key.javaClass.getAnnotation(Listener::class.java)
                val secondAnnotation = o2.key.javaClass.getAnnotation(Listener::class.java)
                if (firstAnnotation == null || secondAnnotation == null) return@sorted 0
                firstAnnotation.priority.compareTo(secondAnnotation.priority)
            }.forEach { entry: Map.Entry<EventListener?, List<Method>> ->
                entry.value.forEach { method ->
                    try {
                        method.invoke(entry.key, eventType)
                    } catch (e: IllegalAccessException) {
                        Conduit.logger.error("Failed to execute plugin event handler method")
                        e.printStackTrace()
                    } catch (e: InvocationTargetException) {
                        Conduit.logger.error("Failed to execute plugin event handler method")
                        e.printStackTrace()
                    }
                }
            }
        }
    }

    companion object {
        @Suppress("UNCHECKED_CAST")
        private fun findEventTypeFromListenerMethod(method: Method): Class<out EventType>? {
            val params = method.parameters
            if (params.isEmpty()) return null // We have to have at least one parameter to check.
            // Check to make sure that this parameter is of the EventType type.
            val param = params[0]
            return if (!EventType::class.java.isAssignableFrom(param.type)) null else param.type as Class<out EventType>?
        }
    }
}
